### PR TITLE
[UNREBASED] Lighter graphs with edge iterators

### PR DESCRIPTION
**Pull request by CarloLucibello**

### CODE REVIEW

**Comment by sbromberger** [comment]
Thanks. The delay in getting to this is because I'm trying to fix LightGraphs so that tests succeed in 0.4 and 0.5, and I'm troubleshooting some DOT parsing issues as well. Stay tuned.

**Comment by sbromberger** [comment]
@CarloLucibello @jpfairbanks a question: since the adjacency lists are vectors, would merging this mean we can have (deterministic, reproducible) edge indices?

File: src/edgeit.jl [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+type EdgeIt
+    m::Int
+    adj::Vector{Vector{Int}}
+    start::EdgeItState # =[it, v, k]
+    directed::Bool
+end
+
+edge_it(g::Graph) = EdgeIt(ne(g), g.fadjlist, EdgeItState(1,1,1), false)
+edge_it(g::DiGraph) = EdgeIt(ne(g), g.fadjlist, EdgeItState(1,1,1), true)
```
Comment (jpfairbanks) [inline src/edgeit.jl]: What about calling this function `edges(g)`?

File: src/edgeit.jl [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+start(eit::EdgeIt) = eit.start
+done(eit::EdgeIt, state::EdgeItState) = state.it > eit.m
+function next(eit::EdgeIt, state::EdgeItState)
+    # println("$eit, $state")
+    assert(state.v <= length(eit.adj))
+    u = 0
+    while state.k <= length(eit.adj[state.v])
```
Comment (jpfairbanks) [inline src/edgeit.jl]: Could you add some comments describing these while loops and why breaking if directed for our future selves?

File: src/edgeit.jl [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        u  = eit.adj[state.v][state.k]
+        eit.directed && break
+        u >= state.v && break
```
Comment (jpfairbanks) [inline src/edgeit.jl]: Is this assuming that adj is not sorted?

Comment (sbromberger) [inline src/edgeit.jl]: let's assume we're going to sort adj lists (#231).

**Comment by sbromberger** [comment]
I'm implementing this in parallel so I can avoid massive rebase issues. @CarloLucibello @jpfairbanks - is there a good way to set a getindex on an iterator? That'd allow us to complete our edge indexing, which may improve other primitives that are using a Dict mapping now.

**Comment by jpfairbanks** [comment]
Do you mean give an iterator random access? I think you would have to turn it into a collection of some kind. edv = collect(edges(g)) will make a vector where edv[i] is the ith edge traversed by the iterator.

The way the edgeiterator is written above, it should be deterministic in that if the graph does not change between two calls, then the edges are visited in the same order.

**Comment by sbromberger** [comment]
So I'm not sure we couldn't do it using the adjacency structure in the EdgeIter object. Basically you just sum up the lengths of each vector until you're > than the index you want, and then calculate the offset in the next-to-last vector you hit.

**Comment by sbromberger** [comment]
Closing this in favor of #237.

**Pull request closed without merging**